Fix wallet address pool gap limited to Word8

hasufell opened the pull request:
Although with Persistent Word32 will change the
underlying type from SqlInt32 to SqlInt64, it
doesn't matter for sqlite, since they all end
up as INTEGER.

https://www.sqlite.org/datatype3.html

Issue #2120

Commit:
Fix wallet address pool gap limited to Word8

Although with Persistent Word32 will change the
underlying type from SqlInt32 to SqlInt64, it
doesn't matter for sqlite, since they all end
up as INTEGER.

https://www.sqlite.org/datatype3.html

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Wallets.hs:
```diff
@@ -525,6 +526,33 @@ spec = do
             verify r expectations
 
+    -- https://github.com/input-output-hk/cardano-wallet/issues/2120
+    it "WALLETS_CREATE_10 - Address pool gap isn't limited to Word8" $ \ctx -> do
```
Would you mind adjusting also test `WALLETS_CREATE_08 - address_pool_gap` to have assertion for `addrPoolMax`... somehow it is missing it.

hasufell replied:
I have a feeling this was omitted precisely of that bug...

hasufell replied:
@piotr-iohk I believe this one is enough then, even and I can delete the other?

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -393,12 +393,12 @@ instance PersistField AddressPoolGap where
 
 instance PersistFieldSql AddressPoolGap where
-    sqlType _ = sqlType (Proxy @Word8)
+    sqlType _ = sqlType (Proxy @Word32)
```
So this corresponds to the newtype

```hs
newtype AddressPoolGap = AddressPoolGap
    { getAddressPoolGap :: Word32 }
    deriving stock (Generic, Show, Eq, Ord)
```

I'm not sure if there is a way to get the inner type of a newtype. Maybe with generics?

KtorZ replied:
```hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}

import Data.Proxy
import Data.Typeable
import Data.Word
import GHC.Generics

newtype AddressPoolGap = AddressPoolGap
    { getAddressPoolGap :: Word32 }
    deriving (Generic, Show)

innerTypeOf :: (Generic a, GInnerType (Rep a)) => a -> Proxy (Inner (Rep a))
innerTypeOf = gInnerType . from

class GInnerType (f :: * -> *) where
    type Inner f :: *
    gInnerType :: f a -> Proxy (Inner f)

instance (GInnerType f) => GInnerType (M1 i c f) where
    type Inner (M1 i c f) = Inner f
    gInnerType (M1 fp) = gInnerType fp

instance GInnerType (K1 i c) where
    type Inner (K1 i c) = c
    gInnerType (K1 c) = Proxy @c


main :: IO ()
main = print $ typeRep $ innerTypeOf $ AddressPoolGap 42
```

```
$ stack runhaskell Main.hs
Word32
```

hasufell replied:
Want it only on type level.

I think this will do:
```hs
type family Inner a where
  Inner (M1 _ _ x) = Inner x
  Inner (K1 _ x) = x
```

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Wallets.hs:
```diff
@@ -462,16 +463,13 @@ spec = do
-                  , [ expectResponseCode @IO HTTP.status201 ]
+                  , [ expectResponseCode @IO HTTP.status201
+                    , expectField (#addressPoolGap . #getApiT) (`shouldBe` maxBound)
+                  ]
```
:+1:

rvl reviewed (approved):
Looks like a fix :+1:.

What about just replacing `WordN` with `Word` in the cases where the range is smallish and doesn't really matter, or where the bounds don't correspond to machine words.

In most cases 4 wasted bytes is nothing, but using specific sized number types distracts readers.

Commit:
Also test listing addresses shows correct number

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Wallets.hs:
```diff
@@ -472,6 +472,14 @@ spec = do
+                , expectListSize 1
+                , expectListField 0 (#addressPoolGap . #getApiT . #getAddressPoolGap)
+                    (`shouldBe` (fromIntegral addrPoolGap))
+                ]
```
I don't think that's what Piotr meant though. He meant to check that listing addresses returns the correct number of addresses.

piotr-iohk replied:
Yes, sorry for not being precise. I meant this guy -> https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/listAddresses. The relationship here is that when we create wallet with address_pool_gap = N, the list address endpoint should return precisely N addresses for that wallet.

hasufell replied:
I think should be fine now

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Wallets.hs:
```diff
@@ -470,8 +470,11 @@ spec = do
+            r@(_, Right wallet) <- request @ApiWallet ctx (Link.postWallet @'Shelley) Default payload
             verify r expectations
+            addrs <- listAddresses @n ctx wallet
+            length addrs `shouldBe` addrPoolGap
```
```suggestion
            rW <- request @ApiWallet ctx (Link.postWallet @'Shelley) Default payload 
            verify rW expectations 
            rA <- listAddresses @n ctx (getFromResponse id r)
            verify rA 
                [ expectListSize addrPoolGap
                ]
```

hasufell committed:
Refactor test to use verify

Co-authored-by: Matthias Benkort <5680256+KtorZ@users.noreply.github.com>

rvl committed:
tests: Prevent conflicts when creating address pool gap test wallet

KtorZ committed:
increase integration HTTP client request timeout to 5 minutes

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Wallets.hs:
```diff
@@ -462,18 +464,23 @@ spec = do
+            rA <- request @[ApiAddress n] ctx
+                (Link.listAddresses @'Shelley w) Default Empty
+            _ <- request @ApiWallet ctx
+                (Link.deleteWallet @'Shelley w) Default Empty
```
Test passes for me locally as well... Could this line be somehow problematic? I guess we don't need to remove the wallet (Just thinking out loud :man_shrugging: )

KtorZ replied:
Rodney added this, because of the hard-coded mnemonic and the automatically retried `it`. If you don't remove it, you get the second run to fail because the wallet already exists :/ ... although it should've been cleaned up in between...

piotr-iohk replied:
oh, ok, haven't noticed that it uses hard-coded mnemonic...

KtorZ committed:
disable newly added assertion requiring listing addresses on max gap wallets
  Added a FIXME here for later. This works locally on 3 different machines, but somehow fails in Buildkite for an unknown
  reason. We should take time investigating this.
